src/components/partnership/partnership.repository.ts

```diff
@@ -265,33 +265,81 @@ export class PartnershipRepository extends DtoRepository(Partnership) {
       .apply(calculateTotalAndPaginateList(Partnership, input));
   }
 
-  async verifyRelationshipEligibility(projectId: ID, partnerId: ID) {
+  async verifyRelationshipEligibility(
+    projectId: ID,
+    partnerId: ID,
+    changeset?: ID
+  ) {
     return (
       (await this.db
         .query()
         .optionalMatch(node('partner', 'Partner', { id: partnerId }))
         .optionalMatch(node('project', 'Project', { id: projectId }))
-        .optionalMatch([
-          node('project'),
-          relation('out', '', 'partnership', { active: true }),
-          node('partnership'),
-          relation('out', '', 'partner', { active: true }),
-          node('partner'),
-        ])
+        .subQuery((sub) =>
+          sub
+            .with('project, partner')
+            .optionalMatch([
+              node('project'),
+              relation('out', '', 'partnership', { active: true }),
+              node('partnership'),
+              relation('out', '', 'partner', { active: true }),
+              node('partner'),
+            ])
+            .return(['partnership'])
+            .apply((q) =>
+              changeset
+                ? q
+                    .union()
```

Comment:
Is a with clause not needed after the union to import the `project, partner` vars again?

Reply:
Honestly, I haven't tested it yet but I don't think it will be needed in the `apply` clause. If the `changeset` is defined, the subquery will be
```
.with('project, partner')
.optionalMatch([
       node('project'),
       relation('out', '', 'partnership', { active: true }),
       node('partnership'),
       relation('out', '', 'partner', { active: true }),
       node('partner'),
])
.return(['partnership'])
.union()
.match([node('changeset', 'Changeset', { id: changeset })])
.optionalMatch([
        node('project'),
        relation('out', '', 'partnership', { active: false }),
        node('partnership'),
        relation('in', '', 'changeset', { active: true }),
        node('changeset'),
 ])
.optionalMatch([
        node('partnership'),
        relation('out', '', 'partner', { active: true }),
        node('partner'),
])
.return(['partnership'])
```

Reply:
[This](https://neo4j.com/developer/kb/post-union-processing/#_post_union_processing_enhancements_in_4_1) has it after the union clause within the subquery. Almost like two separate subqueries being ran independently with their results merged.

I think with the way you have it, those vars don't exist in the unioned part and so they match anything.

I wish there was a way with cypher to be explicit about expecting this var to be previously defined vs creating a new reference.